src/typechecker.erl

```diff
@@ -3438,16 +3438,63 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
+
+-spec check_guard_call(atom(), list()) -> map().
+check_guard_call(is_atom, [{var, _, Var}]) -> #{Var => type(atom)};
+check_guard_call(is_binary, [{var, _, Var}]) -> #{Var => type(binary)};
+check_guard_call(is_bitstring, [{var, _, Var}]) -> #{Var => type(bitstring)};
+check_guard_call(is_boolean, [{var, _, Var}]) -> #{Var => type(boolean)};
+check_guard_call(is_float, [{var, _, Var}]) -> #{Var => type(float)};
+check_guard_call(is_function, [{var, _, Var}]) -> #{Var => type('fun')};
+check_guard_call(is_function, [{var, _, Var}, _]) -> #{Var => type('fun')};
+check_guard_call(is_integer, [{var, _, Var}]) -> #{Var => type(integer)};
+check_guard_call(is_list, [{var, _, Var}]) -> #{Var => type(list)};
+check_guard_call(is_map, [{var, _, Var}]) -> #{Var => type(map)};
+check_guard_call(is_number, [{var, _, Var}]) -> #{Var => type(number)};
+check_guard_call(is_pid, [{var, _, Var}]) -> #{Var => type(pid)};
+check_guard_call(is_port, [{var, _, Var}]) -> #{Var => type(port)};
+check_guard_call(is_record, [{var, _, Var}, {atom, _, Record}]) -> #{Var => type_record(Record)};
+check_guard_call(is_record, [{var, _, Var}, {atom, _, Record}, _]) -> #{Var => type_record(Record)};
+check_guard_call(is_reference, [{var, _, Var}]) -> #{Var => type(reference)};
+check_guard_call(is_tuple, [{var, _, Var}]) -> #{Var => type(tuple)};
+check_guard_call(_Fun, _Vars) -> #{}.
+
+-spec check_guard(#env{}, term()) -> map().
+check_guard(_Env, {call, _, {atom, _, Fun}, Vars}) ->
+    check_guard_call(Fun, Vars);
+check_guard(_Env, {call, _, {remote,_,_,{atom, _, Fun}}, Vars}) ->
+    check_guard_call(Fun, Vars);
+check_guard(Env, {op, _OrElseAnno, 'orelse', Call1, Call2}) ->
+    G1 = check_guard(Env, Call1),
+    G2 = check_guard(Env, Call2),
+    union_var_binds_symmetrical([G1, G2], Env#env.tenv);
+check_guard(Env, {op, _AndAlsoAnno, 'andalso', Call1, Call2}) ->
+    G1 = check_guard(Env, Call1),
+    G2 = check_guard(Env, Call2),
+    union_var_binds([G1, G2], Env#env.tenv);
+check_guard(Env, Guard) ->
+    {_Ty, VB, _Cs} = type_check_expr(Env, Guard), % Do we need to thread the Env?
+    VB.
+
+%% The different guards use glb
+-spec check_guards_sequence(#env{}, list()) -> map().
+check_guards_sequence(Env, GuardSeq) ->
+    RefTys = union_var_binds(
+        lists:map(fun (Guard) ->
+            check_guard(Env, Guard)
+                  end, GuardSeq),
+        Env#env.tenv),
+    maps:merge(Env#env.venv, RefTys).
```

Comment:
This should be the same as for andalso, i.e. `union_var_binds`. Otherwise the last guard overwrites the types refined by previous guards, instead of combining them.

Btw, the terminology according to erlang docs:

* Guard sequences ::= semicolon-separated list of guards
* Guard ::= comma-separated list of guard expressions
* Guard Expression ::= Expression, with limitations

Reply:
Can we come up with a test case that would fail but need to pass? Would love to confirm I fix it.

Reply:
```Erlang
-spec f(number()) -> integer().
f(N) when is_integer(N), is_number(N) -> N;
f(_) -> 42.
```

Reply:
The test passes as it is. I put some printing to make sure it didn't end up being `type(none)` but it was `type(integer)` in the end.

Reply:
Ah... My bad. I see now that `union_var_binds` is used to combine the refined types, just like for `andalso`. Afterwards, `maps:merge/2` is used just to add the refined var binds to then environment. I guess it's fine to let the refined types overwrite the original types, so I was wrong here.

Reply:
Just repeating the terminology, because I think using the proper official terminology is crucial for understanding and maintaining the code.

A comma-separated list of guard expressions is called "guard", so I think the function "check_guards_sequence" should be renamed to "check_guard".

The function "check_guard" should better be renamed to "check_guard_expression".

Also, the variable names used within these functions should better be renamed to reflect this terminology.

(A semicolon-separated list of *guards*  is called "guard sequence", which is handled by "check_guards". The name "check_guards" is fine, I think, but the variable GuardSeq used inside it should be renamed to Guard.)

Reply:
I agree the naming convention should be respected. I will change those names.

Reply:
Renamed. Thank you.